src/server.rs

```diff
@@ -0,0 +1,346 @@
+        self.channel.send(ChannelMessage::Gossip(addr.into()))?;
+        Ok(())
+    }
+}
+
+/// UDP server for ScuttleButt communication.
+struct UdpServer {
+    channel: UnboundedReceiver<ChannelMessage>,
+    scuttlebutt: Arc<Mutex<ScuttleButt>>,
+    socket: Arc<UdpSocket>,
+}
+
+impl UdpServer {
+    async fn new(
+        channel: UnboundedReceiver<ChannelMessage>,
+        scuttlebutt: Arc<Mutex<ScuttleButt>>,
+    ) -> anyhow::Result<Self> {
+        let socket = {
+            let address = &scuttlebutt.lock().await.self_node_id;
+            Arc::new(UdpSocket::bind(address).await?)
+        };
+
+        Ok(Self {
+            scuttlebutt,
+            channel,
+            socket,
+        })
+    }
+
+    /// Listen for new ScuttleButt messages.
+    async fn run(&mut self) -> anyhow::Result<()> {
+        let mut heartbeat_interval = time::interval(MAX_HEARTBEAT_DELTA);
+        let mut gossip_interval = time::interval(GOSSIP_INTERVAL);
+        let mut rng = SmallRng::from_entropy();
+
+        let mut buf = [0; BUF_SIZE];
+        loop {
+            tokio::select! {
+                result = self.socket.recv_from(&mut buf) => match result {
+                    Ok((len, addr)) => {
+                        let _ = self.process_package(addr, &buf[..len]).await;
+                    }
+                    Err(err) => return Err(err.into()),
+                },
+                _ = heartbeat_interval.tick() => self.reap_zombies().await,
+                _ = gossip_interval.tick() => self.gossip_multiple(&mut rng).await,
+                message = self.channel.recv() => match message {
+                    Some(ChannelMessage::Gossip(addr)) => {
+                        let _ = self.gossip(addr).await;
+                    },
+                    Some(ChannelMessage::Shutdown) | None => break,
+                },
+            }
+        }
+        Ok(())
+    }
+
+    /// Process a single UDP packet.
+    async fn process_package(&self, addr: SocketAddr, data: &[u8]) -> anyhow::Result<()> {
+        // Handle gossip from other servers.
+        let message = bincode::deserialize(data)?;
+        let response = self.scuttlebutt.lock().await.process_message(message);
+
+        // Send reply if necessary.
+        if let Some(message) = response {
+            let message = bincode::serialize(&message)?;
+            self.socket.send_to(&message, addr).await?;
+        }
+
+        Ok(())
+    }
+
+    /// Remove all nodes without an up-to-date heartbeat.
+    async fn reap_zombies(&self) {
+        let mut scuttlebutt = self.scuttlebutt.lock().await;
+        let nodes = scuttlebutt.cluster_state_map.node_states();
+        let now = Self::secs_since_epoch();
+
+        nodes.retain(|_node_id, node_state| {
+            let heartbeat = node_state
+                .get(HEARTBEAT_KEY)
+                .and_then(|value| value.parse().ok());
+            now - heartbeat.unwrap_or(0) <= MAX_HEARTBEAT_DELTA.as_secs()
```

Comment:
No. You want to use a monotonous clock to compute timeout.

Reply:
The solution for b) is tricky of course.

The idea is to just use a autoincrement counter as the heartbeat.

The counter in itself is not used.

For a node N to track the liveliness of a given node N' we do need some kind of time information though. 
We get it by calling `Instant::now()` and associating the value to N' whenever we get a sign of liveliness from it:
e.g.:
a) we (the N node) received a message from the N' node
b) we (the N node) received an update about N' from another node N''

Note that in the case of b) the update information could be quite lagging. This is not a problem that hurts the validity of the algorithm. A dead will still be eventually detected as dead.

Reply:
So I'd assume the heartbeat's sole purpose then is to mark a node as "dirty" so nodes will receive an update and update their liveliness status of that node? It's never actually consumed?

And the liveliness tracked in N for N' isn't stored on the N'/Ns k/v storage, but externally, so we don't have that information constantly creating more gossip and create potential feedback-loops?

Reply:
> And the liveliness tracked in N for N' isn't stored on the N'/Ns k/v storage

Yes we don't store that in the gossiped storage.

> It's never actually consumed?

Within the same process, we need a method to list the "live" nodes.

Reply:
> Within the same process, we need a method to list the "live" nodes.

I was talking about the heartbeat stored in the k/v storage in this scenario. That's never "consumed" by any other nodes as far as I understood it.

Reply:
Ah yes. The value is not consumed. The only interesting signal is that it changed.